pocketutils/misc/klgists.py

```diff
@@ -0,0 +1,829 @@
+	return "{}-{}".format(path, timestamp())
+
+
+def nice_time(n_ms: int) -> str:
+	length = datetime.datetime(1, 1, 1) + datetime.timedelta(milliseconds=n_ms)
+	if n_ms < 1000 * 60 * 60 * 24:
+		return "{}h, {}m, {}s".format(length.hour, length.minute, length.second)
+	else:
+		return "{}d, {}h, {}m, {}s".format(length.day, length.hour, length.minute, length.second)
+
+
+def parse_local_iso_datetime(z: str) -> datetime:
+	return datetime.datetime.strptime(z, '%Y-%m-%dT%H:%M:%S.%f')
+
+
+logger = logging.getLogger(__name__)
+
+class PipeType(Enum):
+	STDOUT = 1
+	STDERR = 2
+
+def _disp(out, ell, name):
+	out = out.strip()
+	if '\n' in out:
+		ell(name + ":\n<<=====\n" + out + '\n=====>>')
+	elif len(out) > 0:
+		ell(name + ": <<===== " + out + " =====>>")
+	else:
+		ell(name + ": <no output>")
+	
+
+def _log(out, err, ell):
+	_disp(out, ell, "stdout")
+	_disp(err, ell, "stderr")
+
+	
+def smart_log_callback(source, line, prefix: str = '') -> None:
+	line = line.decode('utf-8')
+	if line.startswith('FATAL:'):
+		logger.fatal(prefix + line)
+	elif line.startswith('ERROR:'):
+		logger.error(prefix + line)
+	elif line.startswith('WARNING:'):
+		logger.warning(prefix + line)
+	elif line.startswith('INFO:'):
+		logger.info(prefix + line)
+	elif line.startswith('DEBUG:'):
+		logger.debug(prefix + line)
+	else:
+		logger.debug(prefix + line)
+
+
+def _reader(pipe_type, pipe, queue):
+	try:
+		with pipe:
+			for line in iter(pipe.readline, b''):
+				queue.put((pipe_type, line))
+	finally:
+		queue.put(None)
+	
+def stream_cmd_call(cmd: List[str], stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell_cmd: str=None, cwd: Optional[str] = None, timeout_secs: Optional[float] = None, log_callback: Callable[[PipeType, bytes], None] = None, bufsize: Optional[int] = None) -> None:
+	"""Calls an external command, waits, and throws a ResourceError for nonzero exit codes.
+	Returns (stdout, stderr).
+	The user can optionally provide a shell to run the command with, e.g. "powershell.exe" 
+	"""
+	if log_callback is None:
+		log_callback = smart_log_callback
+	cmd = [str(p) for p in cmd]
+	if shell_cmd:
+		cmd = [shell_cmd] + cmd
+	logger.debug("Streaming '{}'".format(' '.join(cmd)))
+	
+	p = subprocess.Popen(cmd, stdout=PIPE, stderr=PIPE, cwd=cwd, bufsize=bufsize)
+	try:
+		q = Queue()
+		Thread(target=_reader, args=[PipeType.STDOUT, p.stdout, q]).start()
+		Thread(target=_reader, args=[PipeType.STDERR, p.stderr, q]).start()
+		for _ in range(2):
+			for source, line in iter(q.get, None):
+				log_callback(source, line)
+		exit_code = p.wait(timeout=timeout_secs)
+	finally:
+		p.kill()
+	if exit_code != 0:
+		raise ResourceError("Got nonzero exit code {} from '{}'".format(exit_code, ' '.join(cmd)), cmd, exit_code, '<<unknown>>', '<<unknown>>')
+	
+
+def wrap_cmd_call(cmd: List[str], stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell_cmd: str=None, cwd: Optional[str] = None, timeout_secs: Optional[float] = None) -> (str, str):
```

Comment:
I think this is basically [CallTools `call_cmd_utf`](https://github.com/dmyersturnbull/pocketutils/blob/main/pocketutils/tools/call_tools.py#LL54C6-L54C6).